server/src/main/java/com/linecorp/centraldogma/server/internal/storage/StoragePurgeService.java

```diff
@@ -0,0 +1,164 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.centraldogma.server.internal.storage;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.io.File;
+import java.time.Duration;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableScheduledFuture;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import com.linecorp.centraldogma.common.Author;
+import com.linecorp.centraldogma.server.command.Command;
+import com.linecorp.centraldogma.server.command.CommandExecutor;
+import com.linecorp.centraldogma.server.storage.project.ProjectManager;
+
+import io.netty.util.concurrent.DefaultThreadFactory;
+
+/**
+ * A service class for pursing projects and repositories that were removed before.
+ */
+public class StoragePurgeService {
+
+    private static final Logger logger = LoggerFactory.getLogger(StoragePurgeService.class);
+
+    /**
+     * How often to run the storage purge schedules. i.e. every minute.
+     */
+    private static final Duration TICK = Duration.ofMinutes(1);
+
+    private final ProjectManager projectManager;
+    private final long maxRemovedRepositoryAgeMillis;
+    private final StoragePurgeScheduler storagePurgeScheduler = new StoragePurgeScheduler();
+
+    public StoragePurgeService(
+            ProjectManager projectManager,
+            long maxRemovedRepositoryAgeMillis) {
+        this.projectManager = requireNonNull(projectManager, "projectManager");
+        checkArgument(maxRemovedRepositoryAgeMillis > 0,
+                      "maxRemovedRepositoryAgeMillis: %s (expected: > 0)", maxRemovedRepositoryAgeMillis);
+        this.maxRemovedRepositoryAgeMillis = maxRemovedRepositoryAgeMillis;
+    }
+
+    public void start(CommandExecutor commandExecutor) {
+       storagePurgeScheduler.start(() -> schedulePurgeData(commandExecutor));
+    }
+
+    public void stop() {
+        storagePurgeScheduler.stop();
+    }
+
+    @VisibleForTesting
+    void schedulePurgeData(CommandExecutor commandExecutor) {
+        final long minAllowedTimestamp = System.currentTimeMillis() - maxRemovedRepositoryAgeMillis;
+        final Predicate<File> olderThanMinAllowed = removed -> removed.lastModified() < minAllowedTimestamp;
```

Comment:
Could you use `ProjectMetadata.removal()` and `RepositoryMetadata.removal()` rather than `File.lastModified()`?

Reply:
fixed :-)